src/code/z_eff_spark.c

```diff
@@ -1,9 +1,265 @@
+        this->position = initParams->position;
+        this->speed = initParams->speed;
+        this->gravity = initParams->gravity;
+        this->uDiv = initParams->uDiv;
+        this->vDiv = initParams->vDiv;
+        this->colorStart[0].r = initParams->colorStart[0].r;
+        this->colorStart[0].g = initParams->colorStart[0].g;
+        this->colorStart[0].b = initParams->colorStart[0].b;
+        this->colorStart[0].a = initParams->colorStart[0].a;
+        this->colorStart[1].r = initParams->colorStart[1].r;
+        this->colorStart[1].g = initParams->colorStart[1].g;
+        this->colorStart[1].b = initParams->colorStart[1].b;
+        this->colorStart[1].a = initParams->colorStart[1].a;
+        this->colorStart[2].r = initParams->colorStart[2].r;
+        this->colorStart[2].g = initParams->colorStart[2].g;
+        this->colorStart[2].b = initParams->colorStart[2].b;
+        this->colorStart[2].a = initParams->colorStart[2].a;
+        this->colorStart[3].r = initParams->colorStart[3].r;
+        this->colorStart[3].g = initParams->colorStart[3].g;
+        this->colorStart[3].b = initParams->colorStart[3].b;
+        this->colorStart[3].a = initParams->colorStart[3].a;
+        this->colorEnd[0].r = initParams->colorEnd[0].r;
+        this->colorEnd[0].g = initParams->colorEnd[0].g;
+        this->colorEnd[0].b = initParams->colorEnd[0].b;
+        this->colorEnd[0].a = initParams->colorEnd[0].a;
+        this->colorEnd[1].r = initParams->colorEnd[1].r;
+        this->colorEnd[1].g = initParams->colorEnd[1].g;
+        this->colorEnd[1].b = initParams->colorEnd[1].b;
+        this->colorEnd[1].a = initParams->colorEnd[1].a;
+        this->colorEnd[2].r = initParams->colorEnd[2].r;
+        this->colorEnd[2].g = initParams->colorEnd[2].g;
+        this->colorEnd[2].b = initParams->colorEnd[2].b;
+        this->colorEnd[2].a = initParams->colorEnd[2].a;
+        this->colorEnd[3].r = initParams->colorEnd[3].r;
+        this->colorEnd[3].g = initParams->colorEnd[3].g;
+        this->colorEnd[3].b = initParams->colorEnd[3].b;
+        this->colorEnd[3].a = initParams->colorEnd[3].a;
+        this->duration = initParams->duration;
+
+        this->numElements = (this->uDiv * this->vDiv) + 2;
+        if (this->numElements > ARRAY_COUNT(this->elements)) {
+            return;
+        }
+
+        for (i = 0; i < this->numElements; i++) {
+            EffectSparkElement* elem = &this->elements[i];
+
+            elem->position.x = this->position.x;
+            elem->position.y = this->position.y;
+            elem->position.z = this->position.z;
+            elem->velocity.x = Rand_ZeroOne() - 0.5f;
+            elem->velocity.y = Rand_ZeroOne() - 0.5f;
+            elem->velocity.z = Rand_ZeroOne() - 0.5f;
+
+            velocityNorm = sqrtf(SQ(elem->velocity.x) + SQ(elem->velocity.y) + SQ(elem->velocity.z));
+
+            if (!(fabsf(velocityNorm) < 0.008f)) {
```

Comment:
```suggestion
            if (!IS_ZERO(velocityNorm)) {
```
I'm not sure I like this macro since zero checks are all over the place but the value isn't always `0.008f`, but I suppose it does work here

Reply:
I'm not sure it makes sense to use this macro outside collision stuff since `0.008f` doesn't necessarily work as an epsilon depending on the floats being used, so abstracting it away could lead to issues in some contexts, and it can't be used consistently like you mentioned